DDP-5750 testboston: new enrollment events

yufengwng opened the pull request:
## Context

See DDP-5750. A few new features:

* New event trigger type `USER_STATUS_CHANGED`. Limited to only `ENROLLED` and `COMPLETED` for now.
* New event action type `UPDATE_USER_STATUS`. Limited to only `COMPLETED` for now.
* New PEX expression to query for user's current enrollment status in a study.
* Added support in Housekeeping to handle delayed `UPDATE_USER_STATUS` event actions.
* Added new event configuration for TestBoston "mark completed after 180 days".
* Added custom task `TestBostonCompletedStatusEvent` which we can run in deployed environments to insert new event.

You can test this locally by:
1. Fiddle with event config, so it's 60 secs instead of 180 days.
2. Setup `testboston` study locally.
3. Boot up study-server and Housekeeping locally.
4. Walk through a new participant on local web app.
5. Check logs and database for the COMPLETED status.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.

## FUD Score

- [x] :relaxed: All good, business as usual!

## How do we demo these changes?

- [x] Getting dev into a state where this is user-visible requires some tech fiddling
    - Need to run the new custom task.

## Testing

- [x] I have written automated positive tests
- [ ] I have written automated negative tests
- [x] I have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

(1) Create the new "you're done" SendGrid email template:
```
$ java -Dconfig.file=/path/to/sb.conf -jar target/StudyBuilder.jar testboston \
  --vars /path/to/vars.conf --substitutions studies/testboston/subs.conf \
  --create-emails --email-keys en_study_completed
```

(2) Add the email template id to Vault path `secret/pepper/{env}/v1/testboston/conf`.

(3) Run the custom task to insert new events and run backfill:
```
$ java -Dconfig.file=/path/to/sb.conf -jar target/StudyBuilder.jar testboston \
  --vars /path/to/vars.conf --substitutions studies/testboston/subs.conf \
  --run-task TestBostonCompletedStatusEvent
```

yufengwng committed:
DDP-5750 testboston: new enrollment events

yufengwng commented on study-builder/studies/testboston/study-events.conf:
```diff
@@ -136,6 +136,21 @@
+      },
+      "cancelExpr": """!user.studies["testboston"].isEnrollmentStatus("ENROLLED")""",
+      "maxOccurrencesPerUser": 1,
+      "delaySeconds": 15552000, # 180 days
```
Need confirmation on what this number should be.

yufengwng committed:
DDP-5750 add data backfill

yufengwng commented:
No word about whether we want to move forward with this or not. Closing this PR for now and keeping the branch around.

yufengwng committed:
generalize trigger to USER_STATUS_CHANGE

yufengwng committed:
cleanup

yufengwng commented:
Picking back up on this again. This is ready for another review.

MocanaAtBroad commented:
While going through PR I was confused about the choice of action name to be ENROLLMENT_COMPLETED / EnrollmentCompletedEventAction as it was suggesting to me that it was an event or trigger rather than an operation.
Consider renaming to something that is more suggestive of an operation being executed, like UPDATE_ENROLLMENT_STATUS_TO_COMPLETED / UpdateEnrollmentStatusToCompletedAction (kind of longish but I think more informative). There might be more shorter renames as good or better.

MocanaAtBroad commented:
Also wonder if the action is really just updating the enrollment status, if it could be parameterized so that same action can be used to update to any valid enrollment status in configuration, not just have a hard-coded COMPLETED.

yufengwng committed:
add you're done email

yufengwng committed:
rename and refactor

yufengwng committed:
Merge branch 'develop' into ddp-5750-time-since-enroll

yufengwng committed:
update backfill

yufengwng commented:
@MocanaAtBroad Those are good points! I've renamed and refactored the approach to address those points. However, I am only allowing a limited set of enrollment statuses for now. This way, we can ensure it's safe to allow those status transitions before exposing them as features. Since this is a bit out-of-scope for this PR, if this approach looks good to folks then I can create a separate ticket to expand the enrollment statuses we allow with these new event types.

yufengwng committed:
fix typo
